### ISSUE TITLE
Assignment operator for QPromise

### ISSUE THREAD

**Issue by pwuertz** [simonbrunel/qtpromise#4]
So `QList<T>` appears to be more demanding than expected and requires `<T>` to be assignable as well, oddly enough only for certain build configurations, but still. Although the `std` containers work perfectly fine, for full compatibility with the Qt ecosystem `QPromise` requires `operator=()`.

I actually think that's a good thing because it also allows for building promise chains at runtime without being forced into recursive patterns. E.g:
```cpp
QPromise<void> p = getFirstPromise();
for (int i=0; i < nPromises; ++i)
    p = p.then([]() { /* chain element */ });
return p;
```

So, copy and move assignment support for `QPromise`? If move assignment is supported, shouldn't there be an "invalid/empty" state for `QPromise` too? If there is an empty state, could there be a default constructor creating such a state and curiously bring us back to your other idea of solving the `QVector` issue?

**Comment by simonbrunel** [comment]
> So, copy and move assignment support for QPromise?

Yes 82f48138a9ff13fa39aa55ec1009540731eca4ad

> If there is an empty state, could there be a default constructor ...

I implemented the move assignment as a swap (for `noexcept`) but even if `m_d` was null, I would expect the moved promise to not be accessed anymore. I want to avoid exposing a **public** default constructor because except for containers compatibility, I can't think of any other cases where we should create an "empty" promise?

**Comment by pwuertz** [comment]
Ah, I was about to post a proposal but you beat me to it :).

No, agreed, I don't see a real reason for a default constructor other than satisfying some of the container functionalities. The worst thing that could happen is that a new user unsuccessfully tries to use methods like `QVector::append` or `QList::value` before finding the correct way to achieve a particular goal. The benefit  is, as you say, that users won't be able to create nonsense promises on accident, and of course not being forced to introduce a Null-QPromise and design, code and document how it should behave within the full QtPromise feature set.

Just a small comment concerning move construct and move assign. I think `m_d.reset()` should be called after `swap` in any case. You are right that a moved value "should" not be accessed anymore, but why not enforce it and punish such coding errors. Also, there might be odd cases where someone does `p1 = std::move(p2)` and expects `p1` to be dead rather than living on in `p2` (let `p2` be a global static and the promise encloses some vital resource for extra weirdness).

**Comment by simonbrunel** [comment]
I agree with your comment about the move implementation, but these methods can't be marked as `noexcept` when resetting the shared pointer. I'm not sure that's an issue, but [it seems a good practice](http://www.codingstandard.com/rule/12-5-4-declare-noexcept-the-move-constructor-and-move-assignment-operator/) to declare the MC and MAO as `noexcept`, especially if used with the standard library containers.

Thoughts?

**Comment by simonbrunel** [comment]
Qt are actually doing the same with their `QExplicitlySharedDataPointer` [MAO](https://github.com/qt/qtbase/blob/5.10/src/corelib/tools/qshareddata.h#L206). About the `QPromise` MC, I don't think it's needed to call `other.m_d.reset()` because after the swap, `other` contains a default constructed `m_d` (and thus already [null](https://github.com/qt/qtbase/blob/5.10/src/corelib/tools/qshareddata.h#L164)).

**Comment by pwuertz** [comment]
You're right, this only concerns the MAO. But I dare say that `QExplicitlySharedDataPointer` isn't doing the right thing here, i. e. betrays the programmer about the lifetime of the old object as outlined in the admittedly weird but possible scenario.

The GCC [`std::shared_ptr`](https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/shared_ptr_base.h#L1245) MAO for example constructs a temporary by move, swaps with itself, and is letting its old self destruct at the spot.

About the `noexcept` declaration. As far as I understand it should be perfectly fine to declare a method `noexcept` even when using non-noexcept functions within. You just have to be sure that those functions won't throw or be aware that the application terminates at your method if they do. We know that Qt doesn't use exceptions, so calling `QExplicitlySharedDataPointer::reset` should be safe and any destructor possibly invoked by reset is already `noexcept` by default. Interestingly, `std::shared_ptr::reset` is explicitly defined as`noexcept`, probably because of use cases like that.

**Comment by simonbrunel** [comment]
Make sense, implemented the suggested changes in d3064231599a1b32e3595dd04dd67da7049bd84c
